pyleros/fedec.py

```diff
@@ -0,0 +1,198 @@
+import myhdl
+from myhdl import instances, block, Signal, intbv, \
+                    always_comb, always_seq
+
+from pyleros.types import alu_op_type, t_decSignal, IM_BITS, DM_BITS
+from pyleros.codes import dlist
+
+from pyleros import decoder, rom
+
+
+@block
+def pyleros_fedec(clk, reset, acc, dm_data,
+                pipe_dec, pipe_imme, pipe_dm_addr, pipe_pc, filename=None):
+    """The fedec module for pyleros, that is, the fetch
+    and decode pipeline stage. The modules is purely 
+    combinatorial, except for the updating the pipeline 
+    register. The IM is instantied and only accessed
+    in this stage. The main functions done here are decoding
+    of instruction, setting up branch control signals, selection
+    of other control signals(including the ALU ones), selection 
+    of DM address. an ALU operation on two local variables takes
+    two cycles to execute and another cycle if the result needs 
+    to written back to a local variable
+
+    Arguments (ports):
+        clk: IN Clock signal
+        reset: IN Async reset signal
+        acc: IN Acc value accessed here, not written. This is
+            needed for setting the branch control signals and
+            for memory addredd of JAL
+        dm_data: IN The data read from the DM, which is needed for
+            an direct add or and indirect load/ store(which follows)
+        pipe_dec: OUT List of the decode signals, pass on to the execute stage
+        pipe_imme: OUT Immediate value, as taken from the lower bits 
+                of the instruction, pass on to execute stage
+        pipe_dm_addr: OUT DM read addr, pipeline register
+        pipe_pc: OUT the value of PC, pipeline register
+
+    Parameters:
+
+        filename: Name of the file or a list containing the instructions
+
+    """
+
+    im_addr = Signal(intbv(0)[IM_BITS:])
+
+    instr = Signal(intbv(0)[16:])
+    instr_hi = Signal(intbv(0)[8:])
+
+    branch_en, acc_z = False, True
+
+    pc, pc_next = [Signal(intbv(0)[IM_BITS:])] * 2
+    pc_add = intbv(0)[IM_BITS:]
+
+    decode = [Signal(bool(0)) for i in dlist]
+    decode[int(t_decSignal.op)] = Signal(alu_op_type.LD)
+
+    # Instantiate the instruction memory
+    im_inst = rom.pyleros_im(clk, reset, im_addr, instr, filename=filename)
+
+    # Instantiate the decoder
+    dec_inst = decoder.pyleros_decoder(instr_hi, decode)
+
+    @always_comb
+    def sync_sig():
+
+        # if not reset == reset.active:
+        instr_hi.next = instr[16:8]
+        im_addr.next = pc_next
+        pipe_pc.next = pc_add
+
+
+    @always_comb
+    def dm_addr_sel():
+
+        # if not reset == reset.active:
+        # Note ind ls adds the 8 bit immediate offset
+        # to the previous value of the dm_
+        offset_addr = intbv(dm_data + instr[8:0])[16:]
```

Comment:
Oh my bad (sort of). It will in myhdl, yes, but I don't think it will convert, and you'll have to go through and rework it. Perhaps someone else can comment on the best way to handle interim variables...? Honestly, I think a better way to do this is with an interim signal, which you can assign to in an `always_comb` block. I'm a little suspicious anyway of too much sequential logic in a RTL block.